include/core/rtos/Mutex.hpp

```diff
@@ -0,0 +1,120 @@
+     *
+     * @param waitOption How long (in ticks) the calling thread should wait for the mutex to be available.
+     * Use Enums::TXWait::WaitForever to wait forever.
+     * @return The first error found by the function (or Success if there was no error).
+     */
+    TXError acquire(uint32_t waitOption);
+
+    /**
+     * Releases control of the mutex, allowing other threads to acquire it.
+     *
+     * @return The first error found by the function (or Success if there was no error).
+     */
+    TXError release();
+
+    /**
+     * Prioritizes the list of waiting threads, placing the highest priority waiting thread at the front of the list.
+     * (Does not effect the order of the other waiting threads).
+     *
+     * @return The first error found by the function (or Success if there was no error).
+     */
+    TXError prioritize();
+
+    // Getters
+
+    /**
+     * Retrieves the name of this Mutex.
+     *
+     * @param[out] name a pointer to a place to store the name pointer.
+     * @return The first error found by the function (or Success if there was no error).
+     */
+    TXError getName(char** name);
+
+    /**
+     * Retrieves the ownership count of this mutex. The ownership count is how many times the current
+     * owner of this mutex has locked this mutex. Each time the owner unlocks this mutex, the ownership
+     * count is decremented. Only when ownershipCount reaches 0 does the mutex actually unlock.
+     *
+     * @param[out] ownershipCount a pointer to a place to store the ownership count.
+     * @return The first error found by the function (or Success if there was no error).
+     */
+    TXError getOwnershipCount(uint32_t* ownershipCount);
+
+    /**
+     * Retrieves the name of the thread that currently owns this mutex.
+     *
+     * @param[out] name a pointer to a place to store the name of the owner of the thread.
+     * @return The first error found by the function (or Success if there was no error).
+     */
+    TXError getNameOfOwner(char** ownerName);
+
+    /**
+     * Retrieves the name of the first suspended thread.
+     *
+     * @param[out] threadName a pointer to a place to store the name of the first suspended thread.
+     * @return The first error found by the function (or Success if there was no error).
+     */
+    TXError getNameOfFirstSuspendedThread(char** threadName);
+
+    /**
+     * Retrieves the number of threads that are suspended on this Mutex.
+     *
+     * @param[out] numSuspendedThreads a pointer to a place to store the number of suspended threads.
+     * @return The first error found by the function (or Success if there was no error).
+     */
+    TXError getNumSuspendedThreads(uint32_t* numSuspendedThreads);
+
+private:
+    /**
+     * Threadx struct that actually holds all of the information for the Mutex.
+     */
+    TX_MUTEX txMutex;
+
+    /**
+     * Pointer to the name of the Mutex.
+     */
+    char* name;
```

Comment:
So, I talked about this with Heller, and I agree, but if we were going to do that the smart thing would probably be to copy the name into a allocated array in the BytePool, because then we won't have to have some long fixed-length array that wastes a bunch of space. 

Alternatively, we could just require that the name is always a string literal, because then it is stored in DATA. I'm partial to this solution, because it solves this problem and the problem of getName() returning a pointer at the same time, and it reduces complexity of the code. There are very few times that you would even want to not use a string literal for the name (like say, for child threads), and in those cases they could just use generic names like "child_thread", which would lose a little specificity (you couldn't call it like "child_thread_1", or "child_thread_2"), but it overall is not too big a sacrifice. Additionally, I don't even see where getName would be used except for like log statements.

Now, there is no way to technically enforce that something is a string literal at the code level, BUT, I browsed around stack exchange a little and you can make a pretty simple macro that will cause a compilation error if someone passes in a non-string literal, then we can just also specify it should be a string literal in the doxygen comment.

Let me know if you think that sounds like a good idea.

Reply:
Allocating in the byte pool makes sense, but I also feel like we could get away with limiting the name length to 10 or 20 bytes without much of an issue.

I have 2 issues with alternative. First, we lose the ability to name a thread without a literal. It doesn't sound like a big deal now, but that kind of arbitrary problem can make the RTOS really annoying to work with. We shouldn't make compromises like that unless we have a really good reason for it.

Second, @mjh9585 may disagree, but I really don't like macros. They're very powerful and can do stuff like this, but I think we're liable to end up shooting ourselves in the foot. I'd prefer to avoid them if possible

Reply:
Macros are great, you simply fear their power. Jokes aside, the fixed character array has a lot of potential to waste memory that we might need with all these different thread stacks to ensure nothing is going to crash. I like the idea of enforcing the string literal, but it does limit out ability to debug threads that are dynamically allocated since they would all have the same set of names which is a little bit of a trade off. While it is more complicated, I think copying variables like the name to the beginning of the thread stack is the best option followed by giving the thread the first word after the string (I believe the stack needs to be word aligned).

Reply:
Okay so I was thinking a bit more about this. I think Magee may be right that just having some fixed size for the thread (say, 20 bytes), is probably the most logical thing.
This is because 
1) sure, we're "wasting space" but the thread stack we allocate is magnitudes more space than this, so it's not really a significant overhead.
2) If we go with the other method, where the name is stored in the BytePool, that means we can only store the names once the Bytepool has been initialized, which means the kernel needs to have been started. So we'd need to keep the pointer in the class until the initialize method has been called, then copy the name over, and then update the pointer to point to the new location in the BytePool. This is doable, but _technically_ assumes that the name pointer is valid up until we call init() on the thread. Now, this shouldn't ever be an issue, but if we're worrying about the name falling out of memory we should go all the way.

Reply:
Yeah, implemented fixed-size, but we can change it by redefining a macro.